core/trace/src/main/scala/org/typelevel/otel4s/trace/Tracer.scala

```diff
@@ -161,5 +176,6 @@ object Tracer {
       def noopScope[A](fa: F[A]): F[A] = fa
       def childScope[A](parent: SpanContext)(fa: F[A]): F[A] = fa
       def spanBuilder(name: String): SpanBuilder.Aux[F, Span[F]] = builder
+      def joinOrContinue[A, C: TextMapGetter](carrier: C)(fa: F[A]): F[A] = fa
```

Comment:
I'm coming into this late and without any prior discussion so feel free to link me to prior discussion etc, but would this method be accurate if it was called `rootOrContinue`?  Based on the example above, a general example use-case would be W3 headers come into a service and it needs to figure out if it needs to create a child span of the prior parent or if the span for this service will be the root as there was no context in the carrier (in this case the headers that propagate trace state).

Reply:
In other parts of the library (e.g. `rootScope` or `rootSpan`) the `root` means a brand new span (without a parent) will be created. 

In the current implementation of `joinOrContinue`, if the span details cannot be extracted from the carrier, the given effect `fa` will be executed within the **current** span. In terms of the diagram, it looks like this:

```mermaid
graph LR;
    SpanA[Span A]-->Extract[Extract external 'Span B' from a carrier];
    Extract-->IsExtractedValid{Is extracted span valid?};
    IsExtractedValid --> |Yes| UseExternalSpan[Use 'Span B']
    IsExtractedValid ---->|No| UsePrevious[Keep using 'Span A']
```

For example, if we were operating within the root span and the external span could not be extracted, the processing will continue within the root span. 
____

Then, there are two scenarios:

1) We want to join an external span or start a **root** one.

This should be the most popular case, I believe. To ensure the root scope is a fallback, we do the following:
```scala
Tracer[F].rootScope {
  Tracer[F].joinOrContinue(headers) {
    ???
  }
}
```

2) We want to join an external span in the **middle** of the processing.

Let's say we have some calculations that must be traced once we open a connection, but  **before** we parse the headers. 

Roughly, the flow can be the following:
```
Start processing (root Span A) ->
  Send request, get response ->
    Extract a span from the response headers and join the external span (Span B) -> 
      Create a new span `Span C` within `Span B` ->
        [Optional] Link `Span C` to `Span A` ->
          ???
```

Without linking, the spans are not associated: https://github.com/typelevel/otel4s/pull/125#discussion_r1109488177

Then, to achieve this, we do the following:
```scala
Tracer[F].currentSpanContext.flatMap { current => // local Span A
  Tracer[F].joinOrContinue(headers) { // external Span B
    val builder = Tracer[F].spanBuilder("Span C")
    current.fold(builder)(builder.addLink(_)).build.use { span => // create Span C and link it to Span A
      ???
    }
  }
}
```



_____


After writing this wall of text, I have several conclusions:
1) What is the priority in the method name `rootOrContinue`? If we distinguish the parts of the name into `root` and `continue`, the former feels like a preferred choice and the latter like a fallback. But should it be the opposite? I tend to agree `continueOrRoot` is a better choice.

2) How to handle the scenarios described above properly? Should we have two separate methods? Or one super method with flags, e.g.:
```scala
def tryContinueExternal[F[_], C](carrier: C, linkSpans: Boolean, rootIfNotExtracted: Boolean)(fa: F[A]): F[A]
```

Reply:
The super method with flags looks like boolean blindness.

Does this table seem right?  I don't think the true/true case makes sense, as there's nothing to link to.  And maybe the false/false isn't that useful compared to the two use cases you detailed?

| linkSpans | rootIfNotExtracted | method name          |
|-----------|--------------------|----------------------|
| false     | false              | joinOrContinue       |
| false     | true               | continueOrRoot       |
| true      | false              | joinOrContinueLinked |
| true      | true               | N/A                  |

Reply:
@iRevive I'm not ignoring your well thought out message, just having to do some searching / thinking as to how this is / would be utilized and getting myself acquainted more to have a better more informed opinion.  I got busy with family and Scala Matsuri this weekend :(.

Reply:
@zmccoy no worries, take your time! There is no rush with this PR :)

Reply:
@iRevive Do you have an example of when you'd have an ambient span and then have incoming headers from a service call?  I honestly can't think of any case where this would be something I want with how we've utilized our tracing.  The general use case is:  Service call comes in, we check headers, if they exist we child span from that, otherwise we are root.

Reply:
If I understand you correctly, I think that could come up when processing async messages. You'd have an ambient span representing processing a message coming in over e.g. AMQP. It might have trace headers on the message, but since it's async, you want to link to the incoming span instead of joining the original trace.

Reply:
@zmccoy I spent a bit of time thinking and I cannot come up with a decent use case for `joinOrContinue`. The only more or less relevant scenario is when you are uncertain that the upstream will provide valid headers. And then you would like to continue within the current span. 

But it's a kinda shaky assumption to build the whole method behavior around.

If we look into @bpholt example, the **linking** functionality should be enough. And we already have it in `SpanBuilder`.

Reply:
So I'm hearing we want just `joinOrRoot`, with its current semantics, for now?